syntax = "proto3";

package gymgui.jasonbridge;

import "google/protobuf/timestamp.proto";
import "supervisor.proto";

option cc_enable_arenas = true;

// A percept pushed from Jason environment into gym_gui.
message JasonPercept {
  string name = 1;                    // e.g., "reward_trend"
  string payload_json = 2;            // arbitrary JSON structure
  google.protobuf.Timestamp timestamp = 3; // capture time
}

// Request an action from gym_gui. For now we return a placeholder token
// and leave concrete action mapping for future phases.
message ActionRequest {
  string run_id = 1;                  // current run id
  string agent_id = 2;                // Jason agent identifier
}

message ActionResponse {
  bool available = 1;                 // if an action could be produced
  string action_token = 2;            // symbolic token (future: structured)
  string message = 3;                 // optional diagnostic
}

message Empty {}

// Exposes dedicated RPCs for Jason side.
service JasonBridge {
  // Push one percept asynchronously.
  rpc PushPercept(JasonPercept) returns (gymgui.supervisor.SupervisorControlAck);
  // Apply a supervisor control update.
  rpc ApplyControlUpdate(gymgui.supervisor.SupervisorControlUpdate) returns (gymgui.supervisor.SupervisorControlAck);
  // Request an action (placeholder semantics).
  rpc RequestAction(ActionRequest) returns (ActionResponse);
  // Fetch lightweight supervisor status.
  rpc GetSupervisorStatus(Empty) returns (SupervisorStatus);
}

// Mirrors snapshot() fields; avoids leaking internal details.
message SupervisorStatus {
  bool active = 1;
  bool safety_on = 2;
  string last_action = 3;
  uint32 actions_emitted = 4;
  string last_error = 5;
}
